deepspeed/runtime/zero/stage_1_and_2.py

```diff
@@ -334,18 +335,29 @@ def __init__(self,
             self.round_robin_bit16_groups.append(round_robin_tensors)
             self.round_robin_bit16_indices.append(round_robin_indices)
 
-            # create flat buffer in CPU and move to GPU
-            self.bit16_groups_flat.append(
-                self.flatten_dense_tensors_aligned(
-                    self.round_robin_bit16_groups[i],
-                    self.nccl_start_alignment_factor * dist.get_world_size(group=self.real_dp_process_group[i])).to(
-                        get_accelerator().current_device_name()))
+            # create flat buffer in CPU and move it to the device
+            flattened_buffer = self.flatten_dense_tensors_aligned(
+                self.round_robin_bit16_groups[i],
+                self.nccl_start_alignment_factor * dist.get_world_size(group=self.real_dp_process_group[i]),
+                use_cpu_data=True)
+
+            # free temp CPU params
+            for param in self.bit16_groups[i]:
+                del param.cpu_data
+
+            self.bit16_groups_flat.append(flattened_buffer.to(get_accelerator().current_device_name()))
+            del flattened_buffer
+
+            # recover temp CPU params
+            for param in self.bit16_groups[i]:
```

Comment:
Since the params are already in flattened buffer on the device, I don't understand the need to recover or use `param.cpu_data` beyond this point. I think the `numel()` values in line 360 for padding computation could be derived earlier in line 318 by something like `param.orig_numel = param.numel()`

Reply:
Good point, the cpu_data is required for _update_model_bit16_weights .
which calls _unflatten()_ - This function accept a flat buffer and a list of tensors, where the only use of this list is just to get a corresponding list of shapes, so the flat buffer will be unflattened according to.
At this point the each tensor in _self.round_robin_bit16_groups[group_index]_ has empty storage, only the cpu_data hold the actual shape. 
That is why we had to add _use_cpu_data_ (will be set to True only during initialization), so it will construct the tensor list according to param.cpu_data and not according to param.data.

as a possible solution i can suggest to construct a parallel list of tensors (per group) that will be placed on meta device. then we can get rid of the _orig_shape_.
self.meta_group_tensors = []
....
meta_tensor = torch.tensor(param_shape, dtype=param_dtype, device=meta)
...
self.meta_group_tensors.append(meta_tensor).

and in __update_model_bit16_weights_ we will always use it as an input for _unflatten()_

if you have something else in mind, let me know, i am open for suggestions.

Reply:
Yes, I prefer the idea of meta device since it was created for this kind of situation. Ideally, I would love to modernize the ZeRO code by incorporating more meta device usage, so this is a great start :)

Reply:
Sure, i will have to test it a little bit locally before submitting the change in this PR.

Reply:
@tjruwase - see latest fixes